File: 프로그래머스/2/42586. 기능개발/기능개발.py
from collections import deque


def solution(progresses, speeds):
    q = deque()
    for x, y in zip(progresses, speeds):
        q.append([x, y])

    n = 0
    answer = []
    while q:
        x, y = q.popleft()
        if x >= 100:
            n += 1
            if len(q) == 0:
                answer.append(n)
        elif n != 0:
            answer.append(n)
            x += y
            q.append([x, y])
            n = 0
            for _ in range(len(q) - 1):
                x, y = q.popleft()
                x += y
                q.append([x, y])

        else:
            x += y
            q.append([x, y])
            for _ in range(len(q) - 1):
                x, y = q.popleft()
                x += y
                q.append([x, y])

    return answer

File: 프로그래머스/2/42586. 기능개발/기능개발.java
import java.util.*;
class Solution {
    public int[] solution(int[] progresses, int[] speeds) {
        int N = speeds.length;
        Deque<Integer> dq = new ArrayDeque<>();
        for(int i = 0; i < N; i ++){
            int remain = 100 - progresses[i];
            dq.offer((int)Math.ceil((double)remain/speeds[i]));
        }
        List<Integer> list = new ArrayList<>();
        while(!dq.isEmpty()){
            int cnt = 1;
            int now = dq.poll();
            while(!dq.isEmpty() && now >= dq.peek()){
                dq.poll();
                cnt++;
            }
            list.add(cnt);
        }
        
        return list.stream().mapToInt(i->i).toArray();
    }
}